lib/datasource/git-subdir-commits/index.ts

```diff
@@ -0,0 +1,88 @@
+import crypto from 'crypto';
+import fs from 'fs';
+import simpleGit from 'simple-git';
+import * as packageCache from '../../util/cache/package';
+import { GetReleasesConfig, Release, ReleaseResult } from '../common';
+
+export const id = 'git-subdir-commits';
+
+const cacheMinutes = 10;
+
+// git will prompt for known hosts or passwords, unless we activate BatchMode
+process.env.GIT_SSH_COMMAND = 'ssh -o BatchMode=yes';
+
+export async function getCommits(
+  { lookupName }: GetReleasesConfig,
+  subdir: string
+): Promise<Release[] | null> {
+  let commits = [];
+
+  const cacheNamespace = 'git-subdir-commits';
+  const cachedResult = await packageCache.get<Release[]>(
+    cacheNamespace,
+    lookupName + subdir
+  );
+  /* istanbul ignore next line */
+  if (cachedResult) {
+    return cachedResult;
+  }
+
+  let baseDir;
+  let useCache = false;
+
+  if (
+    process.env.RENOVATE_REPO_CACHE &&
+    fs.existsSync(`${process.env.RENOVATE_REPO_CACHE}/${lookupName}`)
+  ) {
+    useCache = true;
+    baseDir = `${process.env.RENOVATE_REPO_CACHE}/${lookupName}`;
+  } else {
+    baseDir = '/tmp/renovate/' + crypto.randomBytes(20).toString('hex');
+    fs.mkdirSync(baseDir, { recursive: true });
+  }
+
+  try {
+    const git = simpleGit(baseDir);
+    if (!useCache) {
+      await git.clone('git@github.com:' + lookupName, baseDir);
```

Comment:
I think we can avoid using simple-git entirely here, and opt for the github api. We'll avoid storing files on system which should make it more performant and simpler to cache

Reply:
I've looked into it, and GitHub API does allow fetching the log for a specific subdir, which is great, but I'm concerned with the performance of large updates and also with rate-limiting.

Let's say that another Terraform upgrade, syntax update, or wide-scale refactoring touched 50 modules in `infra-modules`.

When we use `RENOVATE_REPO_CACHE`, we do have to store files on the system, and it's not as clean, but `infra-modules` is cloned once and then the log is fetched locally for all subdirs.

With GitHub API, we have to make 50 requests, one per subdir (or one very large request and then implement some subdir-splitting logic on top), which concerns me not even as a performance issues, because 50 requests aren't too many, but as a rate-limiting issue. I don't think Renovate supports retrying requests, so the upgrade will just fail if we get rate-limited.

So yeah, I agree that cloning repos locally is hacky, but it gives us flexibility and large update performance :)

Reply:
Ah, I see what you mean now. You need to be able to fetch each subdir for log support per dir. What I was thinking would be to make a dumb fetch where we get the HEAD of the entire repo, but then you lose subdir information such as the latest change specific to the subdir you're in.